darc/darc.go

```diff
@@ -1180,6 +1191,103 @@ func parseIDEvmContract(in string) (Identity, error) {
 	}, nil
 }
 
+func evalThreshold(visited map[string]bool, getDarc GetDarc,
+	attrFuncs AttrInterpreters, acceptDarc bool, s string, ids []string) error {
+
+	// we have something of form 'threshold<1/2,darc:aa,ed25519:bb,...>'
+
+	s = strings.TrimPrefix(s, "threshold<")
+	s = strings.TrimSuffix(s, ">")
+
+	entries := strings.Split(s, ",")
+
+	if len(entries) < 1 {
+		return xerrors.Errorf("expected at least the threshold fraction: %s", s)
+	}
+
+	fractionStr := entries[0]
+
+	ok := regexp.MustCompile(`\d+/\d+`).MatchString(fractionStr)
+	if !ok {
+		return xerrors.Errorf("the threshold fraction is incorrect: %s", fractionStr)
+	}
+
+	fractionElem := strings.Split(fractionStr, "/")
+	numerator, err := strconv.Atoi(fractionElem[0])
+	if err != nil {
+		return xerrors.Errorf("failed to convert numerator: %v", err)
+	}
+
+	denominator, err := strconv.Atoi(fractionElem[1])
+	if err != nil {
+		return xerrors.Errorf("failed to convert denominator: %v", err)
+	}
+
+	threshold := float64(numerator) / float64(denominator)
+
+	uniqIds := map[string]struct{}{}
+	validIds := map[string]struct{}{}
+
+	for _, entry := range entries[1:] {
+		uniqIds[entry] = struct{}{}
+
+		for _, id := range ids {
+			if id == entry {
+				validIds[id] = struct{}{}
+				break
+			}
+		}
+
+		// In case of a DARC, we need to check the _sign of that DARC and ensure
+		// we don't fall in a loop.
+		if strings.HasPrefix(entry, "darc") {
+			_, found := validIds[entry]
+			if found && acceptDarc {
+				continue
+			}
+
```

Comment:
At this point, either `found == false` or `acceptDarc == false` (or both).
If `found == true` and `acceptDarc == false`, the following recursive validation of DARCs that, if invalid, shortcuts `validIds[entry] = struct{}{}` does not matter because we already have `validIds[entry] == struct{}{}` since `found == true`.

Am I missing something?

Reply:
To be honest I also couldn't understand the logic there. I borrowed the DARC loop check code from above. I was about to remove that piece of code since it seems useless, but I was too afraid I missed something 😅.

Reply:
Yes, I figured it had been borrowed from `evalExprDarc()` :smile: 
But, if I understand correctly, in the above function, the same behavior does not happen, because the "decision" is directly a `return`... Don't know if I am very clear, do you see what I mean?

Reply:
Oh, I see, so yes we should remove that check.

Reply:
Actually I did not mean to remove the check itself (I know I wasn't very clear, sorry).

Per my understanding, the whole point of the `if strings.HasPrefix(entry, "darc") {` test branch is, in case the entry is a DARC, to validate that DARC (possibly recursively, depending on `acceptDarc`). In the original `evalExprDarc()`, the end result of the validation is a boolean in the closure. In your case, it is the inclusion or not of the entry in the `validIds` set. Since you already include the entry in the `validIds` [in the first loop](https://github.com/dedis/cothority/pull/2429/files#diff-3a3df335267a4b1574fe6cb5c8fcace447bb2d612b042e6191ed9e1edcfd8ec2R1237-R1242), even if the recursive validation fails, the entry still remains included. The fix probably involves, like in `evalExprDarc()`, using a boolean variable in the first loop, which would be negated if the validation fails, and the inclusion takes place at the end only if that variable is `true`.

Does this make sense?

Again, this is per my understanding, so it could be totally off. :smile:

Reply:
> Since you already include the entry in the validIds in the first loop, even if the recursive validation fails, the entry still remains included.

Not necessarily. It get included in the first loop only if there is a signed identities of type `darc:...`, but we normally include the identities of the _signed rule.

Reply:
> Not necessarily. It get included in the first loop only if there is a signed identities of type `darc:...`, but we normally include the identities of the _signed rule.
> 
> So at the end the original if is right, I believe. It is a condition the avoid checking the DARC if it is accepted as an identity and present in the list of signed identities.

Right. And if it it not accepted as an identity but is present in the list of signed identities?